### PR TITLE
Implement GCS Filesystem support

### PR DESCRIPTION
**Pull request by xprotobeast2**
Cmake build and FileSystem interface for google cloud storage using C++ client API.

### CODE REVIEW

File: build/CMakeLists.txt [hunk 1]
```diff
@@ -36,6 +36,7 @@ option(TRTIS_ENABLE_CAFFE2 "Include Caffe2 backend in server" OFF)
 option(TRTIS_ENABLE_ONNXRUNTIME "Include ONNXRuntime backend in server" OFF)
 option(TRTIS_ENABLE_PYTORCH "Include PyTorch backend in server" OFF)
 option(TRTIS_ENABLE_CUSTOM "Include Custom backend in server" OFF)
+option(TRTIS_ENABLE_GCS "Include GCS Filesystem support in server" OFF)
```
Comment (deadeyegoodwin) [inline build/CMakeLists.txt]: Move this down to after line 48. This group of enables is for backends.

File: build/CMakeLists.txt [hunk 2]
```diff
@@ -201,6 +202,49 @@ ExternalProject_Add(prometheus-cpp
...
+#
+# Build crc32c
+#
+ExternalProject_Add(crc32c
+  PREFIX crc32c
+  GIT_REPOSITORY "https://github.com/google/crc32c.git"
+  GIT_TAG "1.0.6"
+  SOURCE_DIR "${CMAKE_CURRENT_BINARY_DIR}/crc32c/src/crc32c"
+  CMAKE_CACHE_ARGS
+    -DBUILD_SHARED_LIBS:STRING=yes
```
Comment (deadeyegoodwin) [inline build/CMakeLists.txt]: Can this be built static and have it be built-into GCS library so we don't need to deliver it as separate shared library?

File: build/CMakeLists.txt [hunk 3]
```diff
@@ -201,6 +202,49 @@ ExternalProject_Add(prometheus-cpp
...
+    -DCRC32C_BUILD_TESTS:BOOL=OFF
+    -DCRC32C_BUILD_BENCHMARKS:BOOL=OFF
+    -DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=ON
+    -DCMAKE_INSTALL_PREFIX:PATH=${CMAKE_CURRENT_BINARY_DIR}/crc32c/install
+)
+
+set(CMAKE_PREFIX_PATH ${CMAKE_PREFIX_PATH}
```
Comment (deadeyegoodwin) [inline build/CMakeLists.txt]: Why is this needed? You should be able to use just -DProtobuf_DIR, -DgRPC_DIR and -Dc-ares_DIR in CMAKE_CACHE_ARGS. Doing it this way changes the path for all the projects, not just GCS.

File: build/CMakeLists.txt [hunk 4]
```diff
@@ -201,6 +202,49 @@ ExternalProject_Add(prometheus-cpp
...
+ExternalProject_Add(google-cloud-cpp
+  PREFIX google-cloud-cpp
+  GIT_REPOSITORY "https://github.com/googleapis/google-cloud-cpp.git"
+  GIT_TAG "v0.10.x"
+  SOURCE_DIR "${CMAKE_CURRENT_BINARY_DIR}/google-cloud-cpp/src/google-cloud-cpp"
+  CMAKE_CACHE_ARGS
+    -DGOOGLE_CLOUD_CPP_DEPENDENCY_PROVIDER:STRING=package
+    -DGOOGLE_CLOUD_CPP_PROTOBUF_PROVIDER:STRING=package
+    -DBUILD_TESTING:BOOL=OFF
+    -DBUILD_SHARED_LIBS:BOOL=ON
```
Comment (deadeyegoodwin) [inline build/CMakeLists.txt]: It would be worth trying with static libraries to see what happens... that would save us having to install the extra shared libraries.

File: build/trtis/CMakeLists.txt [hunk 5]
```diff
@@ -31,7 +31,7 @@ if(NOT CMAKE_BUILD_TYPE)
   set(CMAKE_BUILD_TYPE Release)
 endif()
 
-set(CMAKE_CXX_FLAGS "-Wall -Wextra -Wno-unused-parameter -Werror")
+set(CMAKE_CXX_FLAGS "-Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Werror")
```
Comment (deadeyegoodwin) [inline build/trtis/CMakeLists.txt]: We should add this flag only when compiling the files that need it...

File: src/core/filesystem.cc [hunk 6]
```diff
@@ -205,10 +210,317 @@ LocalFileSystem::WriteTextFile(
...
+  Status GetDirectorySubdirs(
+      const std::string& path, std::set<std::string>* subdirs) override;
+  Status GetDirectoryFiles(
+      const std::string& path, std::set<std::string>* files) override;
+  Status ReadTextFile(const std::string& path, std::string* contents) override;
+  Status WriteTextFile(
+      const std::string& path, const std::string& contents) override;
+
+ private:
+  google::cloud::StatusOr<gcs::Client> client_;
```
Comment (deadeyegoodwin) [inline src/core/filesystem.cc]: Put members after functions, separated by blank line (so move to line 238)

File: src/core/filesystem.cc [hunk 7]
```diff
@@ -205,10 +210,317 @@ LocalFileSystem::WriteTextFile(
...
+GCSFileSystem::ParsePath(
+    const std::string& path, std::string* bucket, std::string* object)
+{
+  // Get the bucket name and the object path. Return error if input is malformed
+  size_t bucket_start = path.find("gs://") + 5;
+  size_t bucket_end = path.find("/", bucket_start);
+  *bucket = path.substr(bucket_start, bucket_end - bucket_start);
+  *object = path.substr(bucket_end + 1);
+
+
```
Comment (deadeyegoodwin) [inline src/core/filesystem.cc]: Coding style... don't use multiple blank lines

File: src/core/filesystem.cc [hunk 8]
```diff
@@ -205,10 +210,317 @@ LocalFileSystem::WriteTextFile(
...
+  if (object->empty()) {
+    return Status(
+        RequestStatusCode::INTERNAL, "No file name found in path: " + path);
+  }
+
+  return Status::Success;
+}
+
+std::string
+AppendSlash(std::string& name)
```
Comment (deadeyegoodwin) [inline src/core/filesystem.cc]: const std::string&

File: src/core/filesystem.cc [hunk 9]
```diff
@@ -205,10 +210,317 @@ LocalFileSystem::WriteTextFile(
...
+{
+  // Helper function to take care of lack of trailing slashes
```
Comment (deadeyegoodwin) [inline src/core/filesystem.cc]: Function description goes before the function (line 282)

File: src/core/filesystem.cc [hunk 10]
```diff
@@ -205,10 +210,317 @@ LocalFileSystem::WriteTextFile(
...
+  google::cloud::StatusOr<gcs::BucketMetadata> bucket_metadata =
+      client_->GetBucketMetadata(bucket);
+
+  if (!bucket_metadata) {
+    return Status(
+        RequestStatusCode::INTERNAL,
+        "Could not get MetaData for bucket with name " + bucket);
+  }
+
+  // Check if whether it has children. If at least one child, it is a directory
```
Comment (deadeyegoodwin) [inline src/core/filesystem.cc]: So GCS doesn't allow empty directories?

File: src/core/filesystem.cc [hunk 11]
```diff
@@ -205,10 +210,317 @@ LocalFileSystem::WriteTextFile(
...
+  for (auto&& object_metadata :
+       client_->ListObjects(bucket, gcs::Prefix(AppendSlash(object_path)))) {
+    if (object_metadata) {
+      *is_dir = true;
+      return Status::Success;
```
Comment (deadeyegoodwin) [inline src/core/filesystem.cc]: Use break here instead of having early exit.

File: src/core/filesystem.cc [hunk 12]
```diff
@@ -27,13 +27,16 @@
 #include "src/core/filesystem.h"
 
 #include <dirent.h>
+#include <google/cloud/storage/client.h>
```
Comment (deadeyegoodwin) [inline src/core/filesystem.cc]: Need ifdef TRTIS_ENABLE_GCS to protect the parts of the file that depend on GCS. You can do a build to test and make sure you did it correctly.

**Comment by deadeyegoodwin** [comment]
@xprotobeast2 You also need to rebase to master since there have been recent changes to some common files.

File: src/core/filesystem.cc (hunk 10 above)
Comment (xprotobeast2) [reply src/core/filesystem.cc]: It does but the way I've done it here, empty directories will just have themselves as a child. I added a check to remove these is GetDirectoryContents().

File: src/core/filesystem.cc [hunk 13]
```diff
@@ -205,11 +212,327 @@ LocalFileSystem::WriteTextFile(
...
+}
+
+#endif  // TRTIS_ENABLE_GCS
+
+Status
+GetFileSystem(const std::string& path, FileSystem** file_system)
+{
+#ifdef TRTIS_ENABLE_GCS
+  // Check if this is a GCS path (gs://$BUCKET_NAME)
+  if (!path.empty() && !path.rfind("gs://", 0)) {
```
Comment (deadeyegoodwin) [inline src/core/filesystem.cc]: Here you need to still detect gs:// but if TRTIS_ENABLE_GCS is false then raise an appropriate error about trtis not being compiled with support for gcs.

File: build/CMakeLists.txt [hunk 14]
```diff
@@ -201,6 +204,49 @@ ExternalProject_Add(prometheus-cpp
...
+    -DCRC32C_BUILD_BENCHMARKS:BOOL=OFF
+    -DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=ON
+    -DCMAKE_INSTALL_PREFIX:PATH=${CMAKE_CURRENT_BINARY_DIR}/crc32c
+)
+
+
+# This is necessary because google-cloud-cpp 
+# does not pass protobuf_DIR and gRPC_DIR etc to its 
+# external projects but expects them in CMAKE_PREFIX_PATH
+set(GCS_CMAKE_PREFIX_PATH ${CMAKE_PREFIX_PATH}
```
Comment (deadeyegoodwin) [inline build/CMakeLists.txt]: Can you guard this with TRTIS_ENABLE_GCS
